pkg/controller/certificaterequest/cloudflare.go

```diff
@@ -52,82 +52,74 @@ type CloudflareResponse struct {
-			// Add 5 seconds to ensure Cloudflare's negative
-			// cache record has expired on the next attempt.
-			negativeCacheTTL += 5
-			continue
-		}
-
-		// If there is no answer field, this is likely an expected NXDOMAIN response.
-		if len(response.Answers) == 0 {
-			reqLogger.Info("no answers received from cloudflare; likely not propagated")
-			continue
-		}
-
-		// Trim any trailing dot from the answer name and quotes from the data.
-		cfName := strings.TrimSuffix(response.Answers[0].Name, ".")
-		cfData := strings.Trim(response.Answers[0].Data, "\"")
-
-		if strings.EqualFold(cfName, fqdn) && cfData == txtValue {
-			return true
-		}
-
-		reqLogger.Info("could not validate DNS propagation for " + fqdn)
+	// Set default wait period to 30s
+	waitPeriod := defaultWaitPeriodDNSPropagationCheck
+
+	reqLogger.Info(fmt.Sprintf("attempting to verify resource record %v has been updated with value %v", fqdn, txtValue))
+
+	negativeCacheTTL = 0
+
+	// Fetch DNS records via Cloudflare DNS
+	response, err := FetchResourceRecordUsingCloudflareDNS(reqLogger, fqdn)
+	if err != nil {
+		reqLogger.Error(err, "failed to fetch DNS records")
+		return waitPeriod
+	}
+
+	// Check for a negative cache result and note its TTL.
+	if dnsRCode(response.Status) == dnsRCodeNameError && len(response.Authority) > 0 {
+		negativeCacheTTL = response.Authority[0].TTL
+		reqLogger.Info(fmt.Sprintf("got a negative cache response with a TTL of %v seconds", negativeCacheTTL))
+		// Add 5 seconds to ensure Cloudflare's negative
+		// cache record has expired on the next attempt.
+		negativeCacheTTL += 5
+	}
+
+	// maxNegativeCacheTTL determines what is "reasonable".
+	// If the SOA TTL exceeds this, give up immediately.
+	if negativeCacheTTL > maxNegativeCacheTTL {
+		reqLogger.Info("negative cache TTL is too long; giving up")
+		return maxNegativeCacheTTL
+	}
+
+	// If the TTL is shorter than the default wait time, disregard.
+	if negativeCacheTTL > waitPeriod {
+		waitPeriod = negativeCacheTTL
+	}
+
+	// If there is no answer field, this is likely an expected NXDOMAIN response.
+	if len(response.Answers) == 0 {
+		reqLogger.Info("no answers received from cloudflare; likely not propagated")
+		return waitPeriod
+	}
+
+	// Trim any trailing dot from the answer name and quotes from the data.
+	cfName := strings.TrimSuffix(response.Answers[0].Name, ".")
+	cfData := strings.Trim(response.Answers[0].Data, "\"")
+
+	// Success if returned record value equals expetect txt value
+	if strings.EqualFold(cfName, fqdn) && cfData == txtValue {
+		reqLogger.Info(fmt.Sprintf("record %v has been updated to value %v.", fqdn, txtValue))
+		return 0
 	}
 
-	errMsg := fmt.Sprintf("unable to verify that resource record %v has been updated to value %v after %v attempts.", fqdn, txtValue, maxAttemptsForDnsPropagationCheck)
+	errMsg := fmt.Sprintf("unable to verify that resource record %v has been updated to value %v.", fqdn, txtValue)
 	reqLogger.Error(errors.New(errMsg), errMsg)
```

Comment:
I see that this is carried over but it doesn't feel like we should be returning an error here if we're waiting on caching or the record hasn't propagated yet.

Reply:
Updated to an Info log.